// postgresql-dev-server/download.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
}
Object.defineProperty(exports, "__esModule", { value: true });
const os = __importStar(require("os"));
const http = __importStar(require("https"));
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const child_process = __importStar(require("child_process"));
const process = __importStar(require("process"));
const progress_1 = __importDefault(require("progress"));
const DEFAULT_DOWNLOAD_URLS = {
    linux: 'https://raw.githubusercontent.com/skyjur/node-postgres-dev-server/master/binaries/postgresql-10.3-1-linux-x64-binaries.tar.gz',
    win32: 'https://raw.githubusercontent.com/skyjur/node-postgres-dev-server/master/binaries/postgresql-10.3-1-windows-x64-binaries.zip',
    darwin: 'https://raw.githubusercontent.com/skyjur/node-postgres-dev-server/master/binaries/postgresql-10.3-1-osx-binaries.zip'
};
function getPostgresql(location, user_defined_dowload_urls) {
    var download_urls = Object.assign({}, DEFAULT_DOWNLOAD_URLS, (user_defined_dowload_urls || {}));
    var platform = process.env.PLATFORM || os.platform();
    if (platform in download_urls) {
        var url = download_urls[platform];
        var localPath = path.join(location, path.basename(url));
        return downloadFile(url, localPath)
            .then(() => console.log('Extracting postgresql...'))
            .then(() => extract(localPath, location))
            .catch(console.error);
    }
    else {
        throw new Error(`Have no download link for platform: ${os.platform()}`);
    }
}
function downloadFile(url, localPath) {
    return new Promise((resolve, reject) => {
        if (fs.existsSync(localPath)) {
            return resolve();
        }
        var request = http.get(url, function (response) {
            if (response.statusCode !== 200) {
                reject(new Error('Response code: ' + response.statusCode));
            }
            else {
                var file = fs.createWriteStream(localPath);
                var bar = new progress_1.default(` ${path.basename(url)} [:bar] :rate/bps :percent :etas`, {
                    complete: '=',
                    incomplete: ' ',
                    width: 20,
                    total: parseInt(response.headers['content-length'], 10)
                });
                response.on('data', chunk => bar.tick(chunk.length));
                response.on('error', (err) => {
                    fs.unlink(localPath, () => null);
                    reject(err);
                });
                response.pipe(file);
                file.on('finish', () => {
                    file.end();
                    resolve();
                });
            }
        });
    });
}
function call(command, args) {
    return new Promise((resolve, reject) => {
        var p = child_process.spawn(command, args);
        p.stderr.pipe(process.stderr);
        p.stdout.pipe(process.stdout);
        p.on('close', (code) => {
            if (code > 0) {
                reject(new Error(`Process exit code: ${code}`));
            }
            else {
                resolve();
            }
        });
    });
}
function extract(filepath, location) {
    if (os.platform() == 'win32') {
        return call('PowerShell', ['-Command', `Expand-Archive -Path "${filepath}" -DestinationPath "${location}"`]);
    }
    else if (/tar.gz$/.test(filepath)) {
        return call('tar', ['-x', '-f', filepath, '-C', location, '--overwrite']);
    }
    else if (/zip$/.test(filepath)) {
        return call('unzip', ['-qod', location, filepath]);
    }
    else {
        throw new Error('No rule to extract file');
    }
}
function ensureHaveBinaries(location) {
    if (fs.existsSync(path.join(location, 'pgsql', 'bin'))) {
        return Promise.resolve();
    }
    else {
        return getPostgresql(location);
    }
}
exports.ensureHaveBinaries = ensureHaveBinaries;
if (require.main === module) {
    ensureHaveBinaries(__dirname);
}
